api/all/src/main/java/io/opentelemetry/api/internal/ConfigUtil.java

```diff
@@ -33,10 +34,12 @@ private ConfigUtil() {}
    */
   public static String getString(String key, String defaultValue) {
     String normalizedKey = normalizePropertyKey(key);
+
     String systemProperty =
-        System.getProperties().entrySet().stream()
-            .filter(entry -> normalizedKey.equals(normalizePropertyKey(entry.getKey().toString())))
-            .map(entry -> entry.getValue().toString())
+        System.getProperties().stringPropertyNames().stream()
```

Comment:
`stringPropertyNames` calls `entrySet` without extra synchronization so `ConcurrentModificationException` should still be possible

Reply:
`stringPropertyNames` promises an unmodifiable `Set<String>` based on `enumerateStringProperties()`, which at least in [JDK8](https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/java/util/Properties.java#L1107) seems to be synchronized.

Reply:
Yeah, super weird.
Was also checking the Android implementation, same thing.
https://cs.android.com/android/platform/superproject/main/+/main:libcore/ojluni/src/main/java/java/util/Properties.java

Reply:
does enumerating over `System.getProperties().keys()` help?

Reply:
Couldn't reproduce the `ConcurrentModificationException` on jdk 8, 11 nor 17 even with the original implementation, so maybe `keys()` is not safe.
I'd suggest to create a `clone()` which is marked as synchronized in [jdk 8](https://github.com/openjdk/jdk8u/blob/master/jdk/src/share/classes/java/util/Hashtable.java#L545) and [11](https://github.com/openjdk/jdk11u/blob/master/src/java.base/share/classes/java/util/Properties.java#L1505).

Reply:
Added commit with `clone()`. This did prevent the exception (at least in jdk 8).

Reply:
`clone()` is going to be a lot more expensive

fwiw `keys()` appears to be marked as `synchronized`: https://github.com/openjdk/jdk/blob/9a9add8825a040565051a09010b29b099c2e7d49/jdk/src/share/classes/java/util/Hashtable.java#L256

Reply:
> fwiw keys() appears to be marked as synchronized

I don't think that is enough. https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Hashtable.html says

> The iterators returned by the iterator method of the collections returned by all of this class's "collection view methods" are fail-fast: if the Hashtable is structurally modified at any time after the iterator is created, in any way except through the iterator's own remove method, the iterator will throw a [ConcurrentModificationException](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ConcurrentModificationException.html). Thus, in the face of concurrent modification, the iterator fails quickly and cleanly, rather than risking arbitrary, non-deterministic behavior at an undetermined time in the future. The Enumerations returned by Hashtable's [keys](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Hashtable.html#keys()) and [elements](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Hashtable.html#elements()) methods are not fail-fast; if the Hashtable is structurally modified at any time after the enumeration is created then the results of enumerating are undefined.

Similarly to how you can use iterator from synchronized wrapped created with `Collections.synchronized...` safely by synchronizing on the collection you could do the same here.
```java
Properties properties = System.getProperties();
synchronized (properties) {
  String systemProperty =
      properties.entrySet().stream()
          .filter(entry -> normalizedKey.equals(normalizePropertyKey(entry.getKey().toString())))
          .map(entry -> entry.getValue().toString())
          .findFirst()
          .orElse(null);
  if (systemProperty != null) {
    return systemProperty;
  }
}
```
Alternatively you could reduce the chance of race with making a copy of system properties once.
```java
  @SuppressWarnings({"NullAway", "NonFinalStaticField"})
  private static Map<String, String> systemProperties;
  static {
    initialize();
  }

  // Visible for testing
  static void initialize() {
    systemProperties = System.getProperties().entrySet().stream().collect(
        Collectors.toMap(entry -> normalizePropertyKey(entry.getKey().toString()),
            entry -> entry.getValue().toString()));
  }

  public static String getString(String key, String defaultValue) {
    String normalizedKey = normalizePropertyKey(key);
    String systemProperty = systemProperties.get(normalizedKey);
    if (systemProperty != null) {
      return systemProperty;
    }
    ...
  }
```
with this you'll lose the ability to pick up modifications to system properties. I doubt that this is actually something we need outside of tests, where you can call `initialize` again. Before implementing this approach I'd ask @jack-berg whether this is fine with him.

Reply:
> Alternatively you could reduce the chance of race with making a copy of system properties once.
>
> with this you'll lose the ability to pick up modifications to system properties

if we can scope this one time reading of system properties once per SDK creation that seems ideal (so you can set system properties and then instantiate the SDK)

Reply:
> if we can scope this one time reading of system properties once per SDK creation that seems ideal (so you can set system properties and then instantiate the SDK)

so something like:
```
  private static final Properties systemProperties = (Properties) System.getProperties().clone();
```
??